ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -50,55 +50,44 @@ def __init__(self):
         self.loop()
 
     def loop(self):
-        rate = rospy.Rate(5) # 5Hz
+        rate = rospy.Rate(25) # 25 Hz is half of the twist controller
+        is_lost = True
         while not rospy.is_shutdown():
-            if self.pose is not None and self.waypoints is not None and self.traffic is not None and self.current_velocity is not None:
-                ego_theta = 2.*math.acos(self.pose.orientation.w)
-                ego_vx = np.dot(np.array([[math.cos(ego_theta),-math.sin(ego_theta)],
-                                          [math.sin(ego_theta), math.cos(ego_theta)]]), np.array([1,0]))
-                ego_vy = np.dot(np.array([[0,-1],[1,0]]), ego_vx)
-                #rospy.logerr(ego_theta)
-
-                # publish final_waypoints
-                min_distance = 1e9
-                initial_wp_index = 0
-                for k in xrange(len(self.waypoints)):
-                    i = k
-                    if self.previous_initial_wp_index is not None:
-                        i = (i + self.previous_initial_wp_index) % len(self.waypoints)
-
-                    waypoint = self.waypoints[i].pose.pose.position
-                    np_waypoint = np.array([waypoint.x,waypoint.y])
-                    np_ego_pose = np.array([self.pose.position.x,self.pose.position.y])
-                    np_diff = np_waypoint - np_ego_pose
-                    distance = np.linalg.norm(np_diff)
-                    param = np.dot(np.linalg.inv(np.vstack((ego_vx,ego_vy)).transpose()), np_diff)
-                    # If the waypoint is in front of vehicle and also the closest one,
-                    # update the index.
-                    if(param[0] > 0 and distance < min_distance):
-                        min_distance = distance
-                        initial_wp_index = i % len(self.waypoints)
-                        if self.previous_initial_wp_index is not None:
```

Comment:
This line ensures that the global search is executed only on first loop. From second search, the waypoint index start from the previous one, and only increase until the initial waypoint come to the front of the ego car. So in my opinion, the original code also considers the computational efficiency.

Reply:
Oh nice. I didn't understand at first. Sorry. Now it is clear.
I notice some points in which param[0] behavior are strange, and eventual reverse gear in manual mode are not supported. 
This evening I can compare the output of the 2 methods and the computation time to have more objective evaluation